### DISCUSSION TITLE
Regression in SwiftUI in Xcode 14

### DISCUSSION THREAD

**Discussion by mackoj** [pointfreeco/swift-composable-architecture#1204] [Q&A] [answered]
Hi,

Since WWDC and Xcode 14 I can't compile my project anymore. So I did open a feedback(FB10115879).

> My project compile in Xcode 13.4.1 and in Xcode 14b1 it doesn't, I have "The compiler is unable to type-check this expression in reasonable time; try breaking up the expression into distinct sub-expressions"  error in a few different SwiftUI files that used to compile in Xcode 13.4.1.

**Comment by mackoj** [comment, marked as answer]
Apple did respond this morning I think this could help.

> Engineering has provided the following information regarding this issue:
> 
> We suggest that you refactor views that have WithViewStore call into a separate method:
> 
> Current project:
> 
> public var body: some View { 
> WithViewStore(self.store) { viewStore in 
> {INNER BODY} 
> } 
> } 
> `
> 
> New dedicated body function which accepts ViewStore and moves {INNER BODY} into it:
> 
> public var body: some View { 
> WithViewStore(self.store) { viewStore in 
> body(viewStore: viewStore) 
> } 
> }
> 
> func body(viewStore: ViewStore<{STATE}, {ACTION}>) -> some View { 
> {INNER BODY} 
> }
> 
> Also, please know that the performance problem is related to how result builders work and the fact that WithViewStore is overloaded and has a result builder parameter which forces body of the associated closure to be type-checked multiple times even though some of the overloads might not be applicable. The previous compiler version had an issue where "complex" closures would be allowed to type-check indefinitely which is now fixed and triggers the error in all these cases.
>

**Reply by tgrapperon** [reply]
Hey @mackoj, could you try to build your project with TCA 0.35.0? This is the last version before some additional overloads were added to `WithViewStore`, so if it improves with this version, it could invite to reconsider the changes.
Also, do you have `IfLetStore` or `SwitchStore` in your closure?

**Reply by mackoj** [reply]
Hi @tgrapperon,

The project send in the feedback use TCA 0.35.0 and the places where the issue occur don't use `IfLetStore` or `SwitchStore`.

**Reply by tgrapperon** [reply]
If I agree that extracting block contents into separate variables usually helps, I find the argument about overloading a little weak. There are many built-in views in SwiftUI that are using a `ViewBuilder` argument and which are much more overloaded than `WithViewStore`, for example `Section` or `List`. So maybe it's the `viewStore` argument that is causing issues, even if I find it doubtful.

Assuming you solved your problem by extracting the content into a subview, and if you have time to spend on the issue, you could maybe try to measure the time it takes to build these functions. If this is in the vicinity of 1s or more, you can try to remove a few bits to check when/if it goes down by a lot. This could maybe provide more information about the root of the issue.

**Reply by tgrapperon** [reply]
Alternatively, and for debug purposes, can you try to replace a `WithViewStore` having an issue with this `WithViewStore2`:
```swift
public struct WithViewStore2<State, Action, Content: View>: View {
  private let content: (ViewStore<State, Action>) -> Content
  @ObservedObject private var viewStore: ViewStore<State, Action>

  public init(
    store: Store<State, Action>,
    removeDuplicates isDuplicate: @escaping (State, State) -> Bool,
    content: @escaping (ViewStore<State, Action>) -> Content
  ) {
    self.content = content
    self.viewStore = ViewStore(store, removeDuplicates: isDuplicate)
  }

  public var body: some View {
    content(self.viewStore)
  }
}

extension WithViewStore2 where State: Equatable {
  public init(
    store: Store<State, Action>,
    content: @escaping (ViewStore<State, Action>) -> Content
  ) {
    self = .init(store: store, removeDuplicates: ==, content: content)
  }
}
```
This is a specialized version that has no overloads.
Edit: Edited to remove an internal function that fixes issues in `GeometryReader`s for example, but this shouldn't infer with build performance.

**Reply by mackoj** [reply]
@tgrapperon Thanks, your `WithViewStore2` seems to have fix the compilation issue. 

The feedback was for Xcode 14B1 but Xcode 14B3 brought up the "The compiler is unable to type-check this expression in reasonable time; try breaking up the expression into distinct sub-expressions" in even more places.

**Reply by tgrapperon** [reply]
This is not meant to be used in production, but it shows that there is maybe some more work to do on `WithViewStore` then.

**Reply by tgrapperon** [reply]
My bet is that `WithViewStore`'s `Content` is not locked to be a `View`, so it opens too much possibilities for the content below. Without abusing your patience, does this works in place of a `WithViewStore`? That is, replacing `WithViewStore` by `WithViewViewStore`.
```swift
public struct WithViewViewStore<State, Action, Content: View>: View {
  let content: WithViewStore<State, Action, Content>
  public init(
    store: Store<State, Action>,
    removeDuplicates isDuplicate: @escaping (State, State) -> Bool,
    file: StaticString = #fileID,
    line: UInt = #line,
    content: @escaping (ViewStore<State, Action>) -> Content
  ) {
    self.content = WithViewStore(store: store, removeDuplicates: isDuplicate, content: content)
  }
  
  public init(
    store: Store<State, Action>,
    file: StaticString = #fileID,
    line: UInt = #line,
    content: @escaping (ViewStore<State, Action>) -> Content
  ) where State: Equatable {
    self = WithViewViewStore(store: store, removeDuplicates: ==, content: content)
  }
  
  public var body: some View {
    self.content
  }
}
```

**Reply by tgrapperon** [reply]
Ah, I've seen that you'll likely find a solution with @mbrandonw and @stephencelis. It will be much more effective than trying things like this. Please let me know the outcome!

**Reply by mackoj** [reply]
The `WithViewViewStore` approach works too.

**Reply by tgrapperon** [reply]
Nice! Because it uses a standard `WithViewStore` under the hood. It's maybe possible to write an extension of `WithViewStore` to help with this and that we could trigger using some syntax like
```
WithViewStore(store).asView { … }
```
We would be able to simply add an `.asView` to help the compiler in complex situations. A proper solution is desirable though, and I'm currently experimenting with some `@ObservedStore` property wrapper. I'll report shortly in the form of a discussion.

**Comment by mbrandonw** [comment]
Hey @mackoj, that's a bummer! Is there any chance you could share the project (either publicly or privately with just me and Stephen) so that we could poke around a bit? We would hate if we created compiler problems with some of our latest releases.

**Reply by mackoj** [reply]
What mail should I use to send it privately ?

**Reply by mbrandonw** [reply]
I'll send you an email 😃

**Reply by mbrandonw** [reply]
Also, I should mention, that another workaround besides moving the body of `WithViewStore` to a computed property is to use `@ObservedObject` for your view store:

```swift
struct SomeView: View {
  let store: Store<...>
  @ObservedObject var viewStore: ViewStore<...>

  init(store: ...) { 
    self.store = store 
    self.viewStore = ViewStore(store)
  }

  var body: some View {
    // use self.viewStore directly
  }
}
```

Of course, ideally we fix this compiler issues, but it is just an alternative for now.

**Reply by mackoj** [reply]
Ok thanks I did try this and it compiled too.

**Comment by mackoj** [comment]
FWIW this bug remains in the last Xcode Beta 4

**Reply by stephencelis** [reply]
Thanks for the update! We'll try to take a look at the `WithViewStore` overload situation to see if we can make things work better soon.
